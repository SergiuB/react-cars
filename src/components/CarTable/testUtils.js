export const cars = {
  page1: {
    prevCarUrl: null,
    nextCarUrl: 'page2',
    currentCarUrl: 'page1',
    cars: [{
      id: 1,
      acceleration: 12,
      name: 'Car1',
    }, {
      id: 2,
      acceleration: 15,
      name: 'Car2',
    }]
  },
  page2: {
    prevCarUrl: 'page1',
    nextCarUrl: 'page3',
    currentCarUrl: 'page2',
    cars: [{
      id: 3,
      acceleration: 3,
      name: 'Car3',
    }, {
      id: 4,
      acceleration: 6,
      name: 'Car4',
    }]
  },
  page3: {
    prevCarUrl: 'page2',
    nextCarUrl: null,
    currentCarUrl: 'page3',
    cars: [{
      id: 5,
      acceleration: 36,
      name: 'Car5',
    }]
  }
};

export const api = {
  getCars: url => ({
    then: (f) => {
      f(url ? cars[url] : cars.page1);
    }
  }),
  saveCar: (car) => {
    // console.log('saving car', car);
    const p = { then: f => f() };
    switch (car.id) {
      case 1:
        cars.page1.cars[0] = { ...car };
        return p;
      case 2:
        cars.page1.cars[1] = { ...car };
        return p;
      case 3:
        cars.page2.cars[0] = { ...car };
        return p;
      case 4:
        cars.page2.cars[1] = { ...car };
        return p;
      case 5:
        cars.page3.cars[0] = { ...car };
        return p;
      default:
        return p;
    }
  }
};
