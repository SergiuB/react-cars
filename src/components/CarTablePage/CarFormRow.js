import React from 'react';
import PropTypes from 'prop-types';

import CarForm from '../CarForm';

const CarFormRow = ({ colSpan, ...rest }) => (
  <tr>
    <td colSpan={colSpan}>
      <CarForm {...rest} />
    </td>
  </tr>
);

CarFormRow.propTypes = {
  colSpan: PropTypes.number.isRequired,
};

CarFormRow.defaultProps = {
  colSpan: 1,
};

export default CarFormRow;
