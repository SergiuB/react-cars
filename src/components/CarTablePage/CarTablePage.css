.car-table-page .id-col {
  width: 40px;
}

.car-table-page .large-col {
  width: 120px;
}

.car-table-page .small-col {
  width: 60px;
}