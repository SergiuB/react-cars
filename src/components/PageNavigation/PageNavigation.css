.page-navigation button {
  margin: 5px;
}