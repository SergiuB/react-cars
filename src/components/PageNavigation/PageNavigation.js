const PageNavigation = () => null;

export default PageNavigation;
