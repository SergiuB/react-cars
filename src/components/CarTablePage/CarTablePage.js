import React from 'react';
import PropTypes from 'prop-types';

import CarRow from '../CarRow';
import CarFormRow from './CarFormRow';
import { NONE, carShape } from '../../constants';

const getCarRows = (cars, editingCarId, onRowClick) => cars.map(car => (
  <CarRow
    key={car.id}
    car={car}
    editEnabled={car.id === editingCarId}
    onClick={() => onRowClick(car.id)}
  />
));

const findCarIndexById = (cars, carId) => cars.findIndex(car => car.id === carId);

const CarTablePage = ({
  cars,
  editingCarId,
  onSaveCarChanges,
  onCancelCarChanges,
  onRowClick,
}) => {
  const allRows = getCarRows(cars, editingCarId, onRowClick);

  // insert a CarFormRow for the car to be edited
  if (editingCarId !== NONE) {
    const editingCarIndex = findCarIndexById(cars, editingCarId);
    if (editingCarIndex !== NONE) {
      const car = cars[editingCarIndex];
      allRows.splice(editingCarIndex + 1, 0, (
        <CarFormRow
          key={`${car.id}-form`}
          car={car}
          onSubmit={onSaveCarChanges}
          onCancel={onCancelCarChanges}
        />
      ));
    }
  }

  return (
    <table>
      <thead>
        <tr>
          <th>Id</th>
          <th>Name</th>
          <th>acceleration</th>
        </tr>
      </thead>
      <tbody>
        {allRows}
      </tbody>
    </table>
  );
};

CarTablePage.propTypes = {
  cars: PropTypes.arrayOf(carShape).isRequired,
  editingCarId: PropTypes.number.isRequired,
  onSaveCarChanges: PropTypes.func.isRequired,
  onCancelCarChanges: PropTypes.func.isRequired,
  onRowClick: PropTypes.func.isRequired,
};


CarTablePage.defaultProps = {
  editingCarId: NONE,
  onSaveCarChanges: () => {},
  onCancelCarChanges: () => {},
  onRowClick: () => {},
};

export default CarTablePage;
