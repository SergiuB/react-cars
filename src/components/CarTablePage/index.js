import CarTablePage from './CarTablePage';

export default CarTablePage;
