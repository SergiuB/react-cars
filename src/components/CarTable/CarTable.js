import React, { Component } from 'react';
import PropTypes from 'prop-types';

import CarTablePage from '../CarTablePage';
import PageNavigation from '../PageNavigation';

const tapLog = x => { console.log(x); return x; }

class CarTable extends Component {
  state = {
    cars: [],
    currentCarUrl: null,
    nextCarUrl: null,
    prevCarUrl: null,
    editingCarId: -1,
  }

  componentWillMount() {
    this.updateState();
  }

  getNextCars = () => this.updateState(this.state.nextCarUrl);

  getPrevCars = () => this.updateState(this.state.prevCarUrl);

  updateState = (carUrl, stateToMerge) => this.props.api.getCars(carUrl)
    .then(s => this.setState({ ...s, ...stateToMerge }))

  handleRowClick = cardId =>
    this.setState({ editingCarId: (cardId === this.state.editingCarId) ? -1 : cardId })

  saveCarChanges = car =>
    this.props.api.saveCar(car)
      .then(() => this.updateState(this.state.currentCarUrl, { editingCarId: -1 }))

  render() {
    const {
      cars,
      editingCarId,
      nextCarUrl,
      prevCarUrl,
    } = this.state;

    return (
      <div className='car-table'>
        <CarTablePage
          cars={cars}
          editingCarId={editingCarId}
          onRowClick={this.handleRowClick}
          onSaveCarChanges={this.saveCarChanges}
        />
        <PageNavigation
          nextEnabled={!!nextCarUrl}
          prevEnabled={!!prevCarUrl}
          onNext={this.getNextCars}
          onPrev={this.getPrevCars}
        />
      </div>
    );
  }
}

CarTable.propTypes = {
  api: PropTypes.shape({
    getCars: PropTypes.func,
    saveCar: PropTypes.func,
  }).isRequired,
};

export default CarTable;
