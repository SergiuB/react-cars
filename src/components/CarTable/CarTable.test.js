import React from 'react';
import { shallow } from 'enzyme';

import { CarTable } from './CarTable';

import CarTablePage from '../CarTablePage';
import PageNavigation from '../PageNavigation';

const cars = {
  page1: {
    prevCarUrl: null,
    nextCarUrl: 'page2',
    cars: [{
      id: 1,
      acceleration: 12,
      name: 'Car1',
    }, {
      id: 2,
      acceleration: 15,
      name: 'Car2',
    }]
  },
  page2: {
    prevCarUrl: 'page1',
    nextCarUrl: 'page3',
    cars: [{
      id: 3,
      acceleration: 3,
      name: 'Car3',
    }, {
      id: 4,
      acceleration: 6,
      name: 'Car4',
    }]
  },
  page3: {
    prevCarUrl: 'page2',
    nextCarUrl: null,
    cars: [{
      id: 5,
      acceleration: 36,
      name: 'Car5',
    }]
  }
};

const api = {
  getCars: url => ({
    then: (f) => {
      f(url ? cars[url] : cars.page1);
    }
  }),
  saveCar: (car) => {
    const p = Promise.resolve();
    switch (car.id) {
      case 1:
        cars.page1[0] = { ...car };
        return p;
      case 2:
        cars.page1[1] = { ...car };
        return p;
      case 3:
        cars.page2[0] = { ...car };
        return p;
      case 4:
        cars.page2[1] = { ...car };
        return p;
      case 5:
        cars.page3[0] = { ...car };
        return p;
      default:
        return p;
    }
  }
};

describe('CarTable', function () {
  it('renders correctly with the first page of cars', function () {
    const wrapper = shallow(
      <CarTable
        api={api}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(CarTablePage).prop('cars')).toEqual(cars.page1.cars);
    expect(wrapper.find(PageNavigation).prop('nextEnabled')).toEqual(true);
    expect(wrapper.find(PageNavigation).prop('prevEnabled')).toEqual(false);
  });

  it('renders correctly with the 2nd page of cars', function () {
    const wrapper = shallow(
      <CarTable
        api={api}
      />
    );

    wrapper.find(PageNavigation).prop('onNext')();
    expect(wrapper.find(CarTablePage).prop('cars')).toEqual(cars.page2.cars);
    expect(wrapper.find(PageNavigation).prop('nextEnabled')).toEqual(true);
    expect(wrapper.find(PageNavigation).prop('prevEnabled')).toEqual(true);
  });

  it('renders correctly with the last page of cars', function () {
    const wrapper = shallow(
      <CarTable
        api={api}
      />
    );

    wrapper.find(PageNavigation).prop('onNext')();
    wrapper.find(PageNavigation).prop('onNext')();
    expect(wrapper.find(CarTablePage).prop('cars')).toEqual(cars.page3.cars);
    expect(wrapper.find(PageNavigation).prop('nextEnabled')).toEqual(false);
    expect(wrapper.find(PageNavigation).prop('prevEnabled')).toEqual(true);
  });

  it('sets the currently edited car id according to the clicked row header', function () {
    const wrapper = shallow(
      <CarTable
        api={api}
      />
    );

    wrapper.find(CarTablePage).prop('onRowClick')(2);
    expect(wrapper.find(CarTablePage).prop('editingCardId')).toEqual(2);
  });

  it('resets the currently edited car id if a row header is clicked twice', function () {
    const wrapper = shallow(
      <CarTable
        api={api}
      />
    );

    wrapper.find(CarTablePage).prop('onRowClick')(2);
    wrapper.find(CarTablePage).prop('onRowClick')(2);
    expect(wrapper.find(CarTablePage).prop('editingCardId')).toEqual(-1);
  });

  it('updates the currently edited car id if another clicked row header', function () {
    const wrapper = shallow(
      <CarTable
        api={api}
      />
    );

    wrapper.find(CarTablePage).prop('onRowClick')(2);
    wrapper.find(CarTablePage).prop('onRowClick')(1);
    expect(wrapper.find(CarTablePage).prop('editingCardId')).toEqual(1);
  });
});
