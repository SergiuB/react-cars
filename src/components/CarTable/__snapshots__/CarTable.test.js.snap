// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarTable renders correctly with the first page of cars 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="car-table"
>
    <CarTablePage
        cars={
            Array [
                Object {
                  "acceleration": 12,
                  "id": 1,
                  "name": "Car1",
                },
                Object {
                  "acceleration": 15,
                  "id": 2,
                  "name": "Car2",
                },
              ]
        }
        editingCarId={-1}
        onCancelCarChanges={[Function]}
        onRowClick={[Function]}
        onSaveCarChanges={[Function]}
    />
    <PageNavigation
        nextEnabled={true}
        onNext={[Function]}
        onPrev={[Function]}
        prevEnabled={false}
    />
</div>,
  "nodes": Array [
    <div
      className="car-table"
>
      <CarTablePage
            cars={
                  Array [
                        Object {
                          "acceleration": 12,
                          "id": 1,
                          "name": "Car1",
                        },
                        Object {
                          "acceleration": 15,
                          "id": 2,
                          "name": "Car2",
                        },
                      ]
            }
            editingCarId={-1}
            onCancelCarChanges={[Function]}
            onRowClick={[Function]}
            onSaveCarChanges={[Function]}
      />
      <PageNavigation
            nextEnabled={true}
            onNext={[Function]}
            onPrev={[Function]}
            prevEnabled={false}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <CarTable
        api={
                Object {
                        "getCars": [Function],
                        "saveCar": [Function],
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": CarTable {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "getNextCars": [Function],
        "getPrevCars": [Function],
        "handleRowClick": [Function],
        "props": Object {
          "api": Object {
            "getCars": [Function],
            "saveCar": [Function],
          },
        },
        "refs": Object {},
        "saveCarChanges": [Function],
        "state": Object {
          "cars": Array [
            Object {
              "acceleration": 12,
              "id": 1,
              "name": "Car1",
            },
            Object {
              "acceleration": 15,
              "id": 2,
              "name": "Car2",
            },
          ],
          "currentCarUrl": "page1",
          "editingCarId": -1,
          "nextCarUrl": "page2",
          "prevCarUrl": null,
        },
        "updateState": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="car-table"
>
          <CarTablePage
                    cars={
                              Array [
                                        Object {
                                          "acceleration": 12,
                                          "id": 1,
                                          "name": "Car1",
                                        },
                                        Object {
                                          "acceleration": 15,
                                          "id": 2,
                                          "name": "Car2",
                                        },
                                      ]
                    }
                    editingCarId={-1}
                    onCancelCarChanges={[Function]}
                    onRowClick={[Function]}
                    onSaveCarChanges={[Function]}
          />
          <PageNavigation
                    nextEnabled={true}
                    onNext={[Function]}
                    onPrev={[Function]}
                    prevEnabled={false}
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="car-table"
>
          <CarTablePage
                    cars={
                              Array [
                                        Object {
                                          "acceleration": 12,
                                          "id": 1,
                                          "name": "Car1",
                                        },
                                        Object {
                                          "acceleration": 15,
                                          "id": 2,
                                          "name": "Car2",
                                        },
                                      ]
                    }
                    editingCarId={-1}
                    onCancelCarChanges={[Function]}
                    onRowClick={[Function]}
                    onSaveCarChanges={[Function]}
          />
          <PageNavigation
                    nextEnabled={true}
                    onNext={[Function]}
                    onPrev={[Function]}
                    prevEnabled={false}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarTable
    api={
        Object {
            "getCars": [Function],
            "saveCar": [Function],
          }
    }
/>,
}
`;
